src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.Async.cs

```
@@ -0,0 +1,465 @@
+        // because these are the only two values we need, we only worry about these
+        // if we don't find the Zip64-EOCD, we just give up and try to use the original values
+        if (eocd.NumberOfThisDisk == ZipHelper.Mask16Bit ||
+            eocd.OffsetOfStartOfCentralDirectoryWithRespectToTheStartingDiskNumber == ZipHelper.Mask32Bit ||
+            eocd.NumberOfEntriesInTheCentralDirectory == ZipHelper.Mask16Bit)
+        {
+            // Read Zip64 End of Central Directory Locator
+
+            // This seeks forwards almost to the beginning of the Zip64-EOCDL, one byte after where the signature would be located
+            _archiveStream.Seek(eocdStart - Zip64EndOfCentralDirectoryLocator.SizeOfBlockWithoutSignature, SeekOrigin.Begin);
+
+            // Exactly the previous 4 bytes should contain the Zip64-EOCDL signature
+            // if we don't find it, assume it doesn't exist and use data from normal EOCD
+            if (await ZipHelper.SeekBackwardsToSignatureAsync(_archiveStream,
+                    Zip64EndOfCentralDirectoryLocator.SignatureConstantBytes.AsMemory(),
+                    Zip64EndOfCentralDirectoryLocator.FieldLengths.Signature, cancellationToken).ConfigureAwait(false))
+            {
+                // use locator to get to Zip64-EOCD
+                (bool zip64eocdLocatorProper, Zip64EndOfCentralDirectoryLocator locator) = await Zip64EndOfCentralDirectoryLocator.TryReadBlockAsync(_archiveStream, cancellationToken).ConfigureAwait(false);
+                TryReadZip64EndOfCentralDirectoryPostLocateZip64Block(zip64eocdLocatorProper, locator);
+                // Read Zip64 End of Central Directory Record
+                (bool result, Zip64EndOfCentralDirectoryRecord record) = await Zip64EndOfCentralDirectoryRecord.TryReadBlockAsync(_archiveStream, cancellationToken).ConfigureAwait(false);
+                TryReadZip64EndOfCentralDirectoryPostReadZip64Block(result, record);
```

This method just throws if result is false. Could Zip64EndOfCentralDirectoryRecord.TryReadBlockAsync throw directly, rather than returning a tuple?

Similar principle with Zip64EndOfCentralDirectoryLocator.TryReadBlockAsync, although this result is used in a debug assertion which might actually be better as an exception.

---

Good point. I simplified various of these methods so they return the data directly and do the checks internally instead of deferring them.